src/ffi/logging.rs

```diff
@@ -0,0 +1,236 @@
+    layer::SubscriberExt,
+    registry::LookupSpan,
+};
+
+use crate::{TelioLogLevel, TelioLoggerCb};
+
+pub fn build(log_level: crate::TelioLogLevel, logger: Box<dyn TelioLoggerCb>) -> impl Subscriber {
+    let field_fmt = Delimited::new(
+        " ",
+        debug_fn(|writer, field, value| {
+            if field.name() == "message" {
+                write!(writer, "{value:?}")
+            } else {
+                write!(writer, "{field}={value:?}")
+            }
+        }),
+    );
+
+    tracing_subscriber::registry()
+        .with(LevelFilter::from_level(log_level.into()))
+        .with(
+            fmt::layer()
+                .event_format(TelioEventFmt)
+                .fmt_fields(field_fmt)
+                .with_writer(FfiCallback::new(logger)),
+        )
+}
+
+struct TelioEventFmt;
+
+impl<S, N> FormatEvent<S, N> for TelioEventFmt
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'a> FormatFields<'a> + 'static,
+{
+    fn format_event(
+        &self,
+        ctx: &fmt::FmtContext<'_, S, N>,
+        mut writer: fmt::format::Writer<'_>,
+        event: &tracing::Event<'_>,
+    ) -> std::fmt::Result {
+        let meta = event.metadata();
+        write!(
+            writer,
+            "{:?}:{} ",
+            meta.module_path().unwrap_or("<unknown module>"),
+            meta.line().unwrap_or(0),
+        )?;
+
+        ctx.format_fields(writer.by_ref(), event)?;
+
+        writeln!(writer)
+    }
+}
+
+pub struct FfiCallback {
+    callback: Arc<dyn TelioLoggerCb>,
+}
+impl FfiCallback {
+    fn new(logger: Box<dyn TelioLoggerCb>) -> Self {
+        Self {
+            callback: logger.into(),
+        }
+    }
+}
+
+impl<'a> MakeWriter<'a> for FfiCallback {
+    type Writer = FfiCallbackWriter;
+
+    fn make_writer(&self) -> Self::Writer {
+        unreachable!("make_writer_for is used")
+    }
+
+    fn make_writer_for(&self, meta: &tracing::Metadata<'_>) -> Self::Writer {
+        FfiCallbackWriter {
+            cb: self.callback.clone(),
+            level: (*meta.level()).into(),
+        }
+    }
+}
+
+pub struct FfiCallbackWriter {
+    cb: Arc<dyn TelioLoggerCb>,
+    level: TelioLogLevel,
+}
+
+impl io::Write for FfiCallbackWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // We can be certain that buf fully represents one full event
+        //
+        // See https://github.com/tokio-rs/tracing/blob/tracing-subscriber-0.3.18/tracing-subscriber/src/fmt/fmt_layer.rs#L975
+        //
+        // Additional test added to double protect from future breakages
+
+        let msg = from_utf8(buf)
+            // Trim could be avoided by removing writeln! in formatter
+            // But this makes it more univesal for the future, then we dicide to
+            // switch to other formats
+            .map(|msg| msg.trim().to_string())
+            .unwrap_or_else(|err| format!("failed to decode trace as utf8 bytes: {err:?}"));
```

Comment:
If this function returns an error maybe it would be nice to return it on failed decoding?